src/graph/decremental_connectivity.rs

```diff
@@ -0,0 +1,179 @@
+/// A data-structure that, given a forest, allows dynamic-connectivity queries.
+/// Meaning deletion of an edge (u,v) and checking whether two vertecies are still connected.
+///
+/// # Complexity
+/// The preprocessing phase runs in O(n) time, where n is the the number of vertecies in the forest.
+/// Deletion runs in O(log n) and checking for connectivity runs in O(1) time.
+///
+/// # Sources
+/// used Wikipedia as reference: <https://en.wikipedia.org/wiki/Dynamic_connectivity>
+pub struct DecrementalConnectivity<'a> {
+    adjacent: &'a Vec<Vec<usize>>,
+    component: Vec<usize>,
+    count: usize,
+    visited: Vec<usize>,
+    dfs_id: usize,
+}
+impl<'a> DecrementalConnectivity<'a> {
+    //expects the parent of a root to be itself
+    pub fn new(adjacent: &'a Vec<Vec<usize>>) -> Self {
+        let n = adjacent.len();
+        let mut tmp = DecrementalConnectivity {
+            adjacent,
+            component: vec![0; n],
+            count: 0,
+            visited: vec![0; n],
+            dfs_id: 1,
+        };
+        tmp.component = tmp.calc_component();
+        tmp
+    }
+
+    pub fn connected(&self, u: usize, v: usize) -> Option<bool> {
+        match (self.component.get(u), self.component.get(v)) {
+            (Some(a), Some(b)) => Some(a == b),
+            _ => None,
+        }
+    }
+
+    // original adjacency will not be modified by this function
+    // expects for the graph to have an edge (u,v)
```

Comment:
What do you think about using `panic!` to ensure that?

Reply:
I'd love that, but detecting whether there is an edge between u and v would be in O(n) since, we do not have an adjacency matrix, but rather for each vertex we store the vertices it is adjacent to.

Still looking for a fix

Reply:
In the case where u and v are not immediately adjacent, `delete(u,v)` will still disconnect both vertices in O(log n), by deleting either an edge connected to u, or v. This means, that this function would even then be of value, so maybe just renaming the function to `disconnect(u,v)` would also be an option. 
The only problem i see with this is the ambiguity of the deletion. We can only be certain, that an edge will be deleted and not which edge, but we give the power to the developer to decide, what they want.
What do you think?

Reply:
I am not sure if `disconnect(u,v)` is a good idea. I am simply not sure what what will be the resulting object after such operation.

What do you think about storing all of the edges in some hashset? Correct me if I am wrong, but this will not change the time and space complexity.

Reply:
deletion will now panic, if the edge doesn't exist :)
I've also switched to HashSets, as you recommended